### PR TITLE
29 from csv function

### PR DESCRIPTION
**Pull request by rogerkuou**
Added a `from_csv` function for lazy loading csv files.

- [x] Implement `from_csv`
- [x] Add example dataset
- [x] Unit test
- [x] Documentation
- [x] Exmaple notebook

### CODE REVIEW

File: stmtools/_io.py
Comment (SarahAlidoost) [inline stmtools/_io.py:75]: I think it would be useful for users to know if the pattern in csv files is valid. So perhaps adding `else` statement with a clear error message will help.

File: stmtools/_io.py
Comment (SarahAlidoost) [inline stmtools/_io.py:81]: Can time values be extracted from column names i.e. `amp_20100110` and stored as a time object. If users want to combine datasets, time information is needed.

File: stmtools/_io.py
Comment (SarahAlidoost) [inline stmtools/_io.py:97]: the same comment as above, let's add `else` statement with a clear error message to help users if the pattern in csv file is not valid.

**Review by SarahAlidoost — APPROVED** [review]
@rogerkuou nice work :+1: the function works. Just a suggestion about time coordinates, let's keep them if it is possible.

File: stmtools/_io.py
Comment (fnattino) [inline stmtools/_io.py:91]: I am not sure whether reading a (large) CSV file column by column is a good idea, and I wonder how Dask does it internally. What I am afraid it could happen is that Dask is actually loading the full CSV into memory, or you are reading through all the rows multiple times and each time extract only one value. Probably it is better to read row-oriented formats like CSV files in blocks of rows.

File: stmtools/_io.py
Comment (fnattino) [inline stmtools/_io.py:118]: When creating STMs from SLC files you end up with a Dataset that has time and points dimension, while here you have time and space. Should one maybe use the same notation in both cases? 

Also, the amplitude and phase are treated as coordinates in the SLC case, while they are assigned as variables here, should one treat them in the same way on both sides?

**Review by fnattino — COMMENTED** [review]
Nice @rogerkuou! 

I have just raised two points related to:
* a performance concern I have about reading CSV files by columns;
* making the output of `from_csv` more consistent with the STM produced from SLC data.  

I leave it to you wether you want to consider these now or for later (potential) improvements.

All the rest is very minor stuff (couple of typos and small things related to type hints).

File: stmtools/_io.py
Comment (rogerkuou) [reply stmtools/_io.py:118]: Hi @fnattino, we are working on renaming the dimension name to "space". See #35.

Comment (fnattino) [reply stmtools/_io.py:121]: Ah right, sorry, I got confused. Nice you are already considering renaming 'points' to 'space'.

File: stmtools/_io.py
Comment (rogerkuou) [reply stmtools/_io.py:99]: Good suggestion! I changed the implementation to fist convert the entire DaskDataframe to DaskArray, then index the DaskArray based on the column index

Comment (fnattino) [reply stmtools/_io.py:99]: Uhm, would this change how the data is read in? I think you would still be loading all the dataset in order to extract even a single column

Comment (fnattino) [reply stmtools/_io.py:99]: Performance-wise, I think it would work better if you would read one partition and load all columns from it, then read another partition, and load all columns again, and so on

File: stmtools/_io.py
Comment (rogerkuou) [reply stmtools/_io.py:96]: Thanks for the suggestion! I added on check at the beginning of the function, to check if all specified space/space-time patterns have at least one match. Otherwise an ValueError is raised

File: stmtools/_io.py
Comment (rogerkuou) [reply stmtools/_io.py:102]: I will tentatively hold this for now. Because currently the time information is all string-based (e.g. from folder names and col headers, etc.) This can be fragile and may be even incorrect in case of futural hourly observation. We are thinking of integrating the time information from metadata, e.g. STAC-Catalog metatdata, so for now I will leave it to users to manually change the time coordinates.

However I think maybe we should do a good documentation on this. Do you have any suggestion?

File: stmtools/_io.py
Comment (rogerkuou) [reply stmtools/_io.py:112]: Ah I see what you mean. That make sense. I will think of a way how to do this

Comment (rogerkuou) [reply stmtools/_io.py:112]: Hi @fnattino, I inspected the following reading resullts:

```python
data = stmtools.from_csv(p_data, blocksize=100e6)
data
```

```
<xarray.Dataset>
Dimensions:                (space: 50000, time: 198)
Coordinates:
  * space                  (space) int64 0 1 2 3 4 ... 49996 49997 49998 49999
  * time                   (time) int64 0 1 2 3 4 5 ... 192 193 194 195 196 197
    lat                    (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    lon                    (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
Data variables: (12/25)
    pnt_id                 (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    pnt_flags              (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    ...                     ...
    pnt_std_linear         (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    pnt_std_quadratic      (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    pnt_std_seasonal       (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    deformation            (space, time) object dask.array<chunksize=(25000, 198), meta=np.ndarray>
    amplitude              (space, time) object dask.array<chunksize=(25000, 198), meta=np.ndarray>
    h2ph                   (space, time) object dask.array<chunksize=(25000, 198), meta=np.ndarray>
```

The data has a `space` size of 50000 and has a chunk size `25000`, so it should be chunked into 2 parts in space.  I visualized two dask graphs of the loaded Dataset, focusing on 1 and 2 data variables:

```python
# Select some vars and remove coords to simplify
data2_one_var = data[['pnt_id']].drop(['lat', 'lon'])
data2_two_var = data[['pnt_id', 'pnt_flags']].drop(['lat', 'lon'])
```

Graph for 1 variable:
dask_1var

Graph for 2 variables: 
dask_2var

I interpret the two graphs as Dask is already doing what we want it to do: first loading each row (space) chunk, and within each chunk get the column attribute. Therefore I would suggest we leave the implementation like this? I will add some documentation to doc this behavior of Dask.

Comment (fnattino) [reply stmtools/_io.py:112]: Sounds good! Also, we can test how the function performs, and if we see it's becoming a bottleneck we can have a look at it again in future!

**Pull request merged**
